server/license: Initial support for serverless

spilchen opened the pull request:
Previously, the license enforcer was not initialized for secondary tenants. There are two modes for secondary tenants: when the tenant runs as a separate process from the system tenant (serverless), and when it shares the same process. In the shared process mode, the enforcer relied on a shared singleton—initialized for the system tenant and reused for secondary tenants. However, when the secondary tenant runs in a separate process (serverless), the enforcer had throttling fully disabled.

This change is the first step in supporting serverless. The main challenge in allowing secondary tenants to initialize the enforcer is that they don’t have access to the KV key stored in the system keyspace, which records the grace period's end when no license is installed. This change doesn’t resolve that yet, but it sets the foundation for future work. For now, it estimates the grace period by setting it to 7 days from the time the enforcer is created.

Several changes were made to support serverless in this form:
- Call the enforcer’s `Start()` function for secondary tenants as well.
- Allow `Start()` to be called multiple times.
- Move all parameters for `Start` into an options struct.
- Remove the enforcer singleton, as it caused more issues (especially in tests) than benefits.
- Secondary tenants that share the same process will still share the same enforcer, but now the enforcer is passed around by storing a copy in `SQLConfig`.

This change will be backported to 24.2, 24.1, 23.2 and 23.1.

Epic: CRDB-39988
Informs: CRDB-42309
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131347)
<!-- Reviewable:end -->

spilchen committed:
server/license: Initial support for serverless

Previously, the license enforcer was not initialized for secondary tenants.
There are two modes for secondary tenants: when the tenant runs as a separate
process from the system tenant (serverless), and when it shares the same
process. In the shared process mode, the enforcer relied on a shared
singleton—initialized for the system tenant and reused for secondary tenants.
However, when the secondary tenant runs in a separate process (serverless), the
enforcer had throttling fully disabled.

This change is the first step in supporting serverless. The main challenge in
allowing secondary tenants to initialize the enforcer is that they don’t have
access to the KV key stored in the system keyspace, which records the grace
period's end when no license is installed. This change doesn’t resolve that
yet, but it sets the foundation for future work. For now, it estimates the
grace period by setting it to 7 days from the time the enforcer is created.

Several changes were made to support serverless in this form:
- Call the enforcer’s `Start()` function for secondary tenants as well.
- Allow `Start()` to be called multiple times.
- Move all parameters for `Start` into an options struct.
- Remove the enforcer singleton, as it caused more issues (especially in tests)
  than benefits.
- Secondary tenants that share the same process will still share the same
  enforcer, but now the enforcer is passed around by storing a copy in
  `SQLConfig`.

Epic: CRDB-39988
Informs: CRDB-42309
Release note: None

fqazi reviewed (commented):
> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131347)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @spilchen, and @vidit-bhat)

___
*[`pkg/server/license/enforcer.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-O7dYLdm6ENj4RVzBMWB:-O7dYLdm6ENj4RVzBMWC:b-qwlucy) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d3b9b5224d8fab8901ee4ec5b500bf1e4eeec/pkg/server/license/enforcer.go#L49)):*
> ```Go
> 
> 	// testingKnobs are used to control the behavior of the enforcer for testing.
> 	testingKnobs *TestingKnobs
> ```

Did you want to move this outside of the mutex?
___
*[`pkg/server/license/enforcer.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-O7dYk5j57AXnpYsC-zb:-O7dYk5j57AXnpYsC-zc:b953ogv) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d3b9b5224d8fab8901ee4ec5b500bf1e4eeec/pkg/server/license/enforcer.go#L222)):*
> ```Go
> 		gracePeriodLength := e.getGracePeriodDuration(7 * 24 * time.Hour)
> 		end := e.getStartTime().Add(gracePeriodLength)
> 		e.clusterInitGracePeriodEndTS.Store(end.Unix())
> ```

Does this mean this time could be reset by restarting the tenant?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131347)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @herkolategan, and @vidit-bhat)

___
*[`pkg/server/license/enforcer.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-O7dYLdm6ENj4RVzBMWB:-O7e0rgw6Cm9NEYm635N:b-wsrmp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d3b9b5224d8fab8901ee4ec5b500bf1e4eeec/pkg/server/license/enforcer.go#L49)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Did you want to move this outside of the mutex?
</blockquote></details>

Yes, this was intentional. I can now set/get the testing knobs without a mutex because the license enforcer is no longer a singleton. Previously, tests that started a test server in parallel would share the same enforcer, leading to race conditions. Now, each test creates its own enforcer, eliminating those issues.
___
*[`pkg/server/license/enforcer.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-O7dYk5j57AXnpYsC-zb:-O7e1_X0BJ3i4BhpYVps:b-vbd7ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d3b9b5224d8fab8901ee4ec5b500bf1e4eeec/pkg/server/license/enforcer.go#L222)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Does this mean this time could be reset by restarting the tenant?
</blockquote></details>

Good question. You’re right—restarting the node resets the grace period for clusters without a license. It’s not ideal, but it's an improvement over the previous state. Before, the enforcer for serverless clusters had throttling entirely disabled. As a follow-up, I hope to allow the secondary tenant to read the KV key as an exception.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-:-O7eB2pM0gSZoxwn3XG4:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @spilchen, and @vidit-bhat)

___
*[`pkg/server/license/enforcer.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131347#-O7dYk5j57AXnpYsC-zb:-O7eB-lpAPJjgkTm5uJL:bk0ezuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/101d3b9b5224d8fab8901ee4ec5b500bf1e4eeec/pkg/server/license/enforcer.go#L222)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good question. You’re right—restarting the node resets the grace period for clusters without a license. It’s not ideal, but it's an improvement over the previous state. Before, the enforcer for serverless clusters had throttling entirely disabled. As a follow-up, I hope to allow the secondary tenant to read the KV key as an exception.
</blockquote></details>

Ah, makes sense.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @fqazi 

bors r=fqazi
